From cd11404e647092dd1eaf87f2fefb8b810bdd6960 Mon Sep 17 00:00:00 2001
From: Gon Solo <gonsolo@gmail.com>
Date: Thu, 10 Oct 2024 09:30:19 +0200
Subject: [PATCH] Fix firemarshal build.

---
 wlutil/wlutil.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/wlutil/wlutil.py b/wlutil/wlutil.py
index 7d30eda6b5..cf74688415 100644
--- a/wlutil/wlutil.py
+++ b/wlutil/wlutil.py
@@ -595,7 +595,7 @@ def mountImg(imgPath, mntPath):
 
     assert imgPath.is_file(), f"Unable to find {imgPath} to mount"
     ret = run(["mountpoint", mntPath], check=False).returncode
-    assert ret == 1, f"{mntPath} already mounted. Somethings wrong"
+    assert ret != 0, f"{mntPath} already mounted. Somethings wrong"
 
     uid = sp.run(['id', '-u'], capture_output=True, text=True).stdout.strip()
     gid = sp.run(['id', '-g'], capture_output=True, text=True).stdout.strip()
-- 
2.47.0

